File: ts/src/api_demo.ts
import "dotenv/config";

import fs from "fs/promises";
import path from "path";
import { fileURLToPath } from "url";
import { LokaliseApi } from "@lokalise/node-api";
import AdmZip from "adm-zip";

async function main() {
  const lokaliseApi = new LokaliseApi({ apiKey: process.env.API_KEY });

  console.log("Creating project...");

  const project = await lokaliseApi.projects().create({
    name: "Node.js Sample Project",
    description: "Here's my Node.js project",
    languages: [
      {
        lang_iso: "en",
      },
      {
        lang_iso: "fr",
      },
    ],
    base_lang_iso: "en",
  });

  const projectId = project.project_id;

  console.log(projectId);
  console.log(project.name);

  // // If needed, provide the project ID manually:
  // const projectId = "41927157619e6abd190863.11993227";

  console.log("Inviting contributors...");

  const contributors = await lokaliseApi.contributors().create(
    [
      {
        email: "translator@example.com",
        fullname: "Mr. Translator",
        is_admin: false,
        is_reviewer: true,
        languages: [
          {
            lang_iso: "en",
            is_writable: false,
          },
          {
            lang_iso: "fr",
            is_writable: true,
          },
        ],
      },
    ],
    { project_id: projectId },
  );

  console.log(contributors[0].email);
  console.log(contributors[0].user_id);

  console.log("Uploading translations...");

  const __filename = fileURLToPath(import.meta.url);
  const __dirname = path.dirname(__filename);
  const i18nFolder = path.resolve(__dirname, "i18n");
  const i18nFile = path.join(i18nFolder, "en.json");
  const data = await fs.readFile(i18nFile, "utf8");
  const buff = Buffer.from(data, "utf8");
  const base64I18n = buff.toString("base64");

  const bgProcess = await lokaliseApi.files().upload(projectId, {
    data: base64I18n,
    filename: "en.json",
    lang_iso: "en",
  });

  console.log("Updating process status...");

  await waitUntilUploadingDone(lokaliseApi, bgProcess.process_id, projectId);

  console.log("Uploading is done!");

  console.log("Getting created translation keys...");

  const keys = await lokaliseApi.keys().list({
    project_id: projectId,
  });
  const keyIds = keys.items.map((currentValue) => currentValue.key_id);

  console.log(keyIds);

  console.log("Assinging a translation task...");

  const task = await lokaliseApi.tasks().create(
    {
      title: "Translate French",
      keys: keyIds, // use ids obtained on the previous step
      languages: [
        {
          language_iso: "fr",
          users: [contributors[0].user_id], // an array of task assignee, we add the previously invited user
        },
      ],
    },
    { project_id: projectId },
  );

  console.log(task.title);
  console.log(task.languages[0].language_iso);

  console.log("Downloading translations...");

  const downloadResponse = await lokaliseApi.files().download(projectId, {
    format: "json",
    original_filenames: true,
    directory_prefix: "",
    filter_langs: ["fr"],
    indentation: "2sp",
  });
  const translationsUrl = downloadResponse.bundle_url;
  const zip = new AdmZip(await zipBuffer(translationsUrl));
  zip.extractAllTo(i18nFolder, true);
}

async function zipBuffer(translationsUrl: string): Promise<Buffer> {
	const response = await fetch(translationsUrl);
	const arrayBuffer = await response.arrayBuffer();

	return Buffer.from(new Uint8Array(arrayBuffer));
}

async function waitUntilUploadingDone(
	lokaliseApi: LokaliseApi,
	processId: string,
	projectId: string,
): Promise<string> {
	return new Promise<string>((resolve, reject) => {
		const interval = setInterval(async () => {
			try {
				const reloadedProcess = await lokaliseApi
					.queuedProcesses()
					.get(processId, {
						project_id: projectId,
					});

				if (reloadedProcess.status === "finished") {
					clearInterval(interval);
					resolve(reloadedProcess.status);
				}
			} catch (error) {
				clearInterval(interval);
				console.error("An error occurred:", error);
				reject("error");
			}
		}, 1000);
	});
}

main()
	.then(() => process.exit(0))
	.catch((error) => {
		console.error(error);
		process.exit(1);
	});

File: python/src/api_demo.py
import lokalise
import requests
import io
import zipfile
import os
import base64
import time

from dotenv import load_dotenv

load_dotenv()


client = lokalise.Client(os.getenv('API_KEY'))

print("Creating a new project...")

project = client.create_project({
    "name": "Python Sample Project",
    "description": "Here's my Python project",
    "languages": [
        {"lang_iso": "en"},
        {"lang_iso": "fr"}
    ],
    "base_lang_iso": "en"
})

project_id = project.project_id

print("Inviting contributors...")

contributors = client.create_contributors(project_id, [
    {
        "email": "translator@example.com",
        "fullname": "Mr. Translator",
        "is_admin": False,
        "is_reviewer": True,
        "languages": [
            {
                "lang_iso": "en",
                "is_writable": False
            },
            {
                "lang_iso": "fr",
                "is_writable": True
            }
        ]
    }
])

contributor = contributors.items[0]

print(contributor.email)
print(contributor.user_id)

print("Uploading translation file...")


def is_uploaded(api_client, project, process):
    for _i in range(5):
        process = api_client.queued_process(project, process.process_id)

        if process.status == 'finished':
            return True

        time.sleep(1)

    return False


filename = os.path.join(os.path.dirname(__file__), 'i18n/en.json')
with open(filename) as f:
    content = f.read()
    file_data = base64.b64encode(content.encode()).decode()

    bg_process = client.upload_file(project_id, {
        "data": file_data,
        "filename": 'en.json',
        "lang_iso": 'en'
    })

    print(f"Checking status for process {bg_process.process_id}...")
    result = is_uploaded(client, project_id, bg_process)
    print(result)

print("Fetching translation keys...")

keys = client.keys(project_id).items
key_ids = list(map(lambda k: k.key_id, keys))
print(key_ids)

print("Assigning a translation task...")

task = client.create_task(project_id, {
    "title": "Translate French",
    "keys": key_ids,
    "languages": [
        {
            "language_iso": "fr",
            "users": [contributor.user_id]
        }
    ]
})

print(task.title)
print(task.languages[0].language_iso)

print("Downloading translation file...")

response = client.download_files(project_id, {
    "format": "json",
    "filter_langs": ["fr"],
    "original_filenames": True,
    "directory_prefix": ""
})

data = io.BytesIO(requests.get(response['bundle_url']).content)

with zipfile.ZipFile(data) as archive:
    archive.extract("fr.json", path="src/i18n/")